driver/kubernetes/factory.go

```diff
@@ -107,6 +133,45 @@ func (f *factory) New(ctx context.Context, cfg driver.InitConfig) (driver.Driver
 	return d, nil
 }
 
+func (f *factory) getClientConfig(endpoint string, dockerCli command.Cli) ClientConfig {
+	if f.cc != nil {
+		return f.cc
+	}
+
+	contextStore := dockerCli.ContextStore()
+
+	var err error
+	var cc ClientConfig
+	cc, err = ctxkube.ConfigFromEndpoint(endpoint, contextStore)
+	if err != nil {
+		// err is returned if n.Endpoint is non-context name like "unix:///var/run/docker.sock".
+		// try again with name="default".
+		// FIXME(@AkihiroSuda): n should retain real context name.
+		cc, err = ctxkube.ConfigFromEndpoint("default", contextStore)
+		if err != nil {
+			logrus.Error(err)
+		}
+	}
+
+	tryToUseKubeConfigInCluster := false
+	if cc == nil {
+		tryToUseKubeConfigInCluster = true
+	} else {
+		if _, err := cc.ClientConfig(); err != nil {
+			tryToUseKubeConfigInCluster = true
+		}
+	}
+	if tryToUseKubeConfigInCluster {
+		ccInCluster := ClientConfigInCluster{}
+		if _, err := ccInCluster.ClientConfig(); err == nil {
+			logrus.Debug("using kube config in cluster")
+			cc = ccInCluster
+		}
+	}
```

Comment:
didn't touch this logic from builder package yet, should look at it more closely as follow-up.